src/Runner.Listener/JobDispatcher.cs

```diff
@@ -675,9 +685,128 @@ private async Task RunAsync(Pipelines.AgentJobRequestMessage message, string orc
             }
         }
 
-        public async Task RenewJobRequestAsync(int poolId, long requestId, Guid lockToken, string orchestrationId, TaskCompletionSource<int> firstJobRequestRenewed, CancellationToken token)
+        internal async Task RenewJobRequestAsync(Pipelines.AgentJobRequestMessage message, ServiceEndpoint systemConnection, int poolId, long requestId, Guid lockToken, string orchestrationId, TaskCompletionSource<int> firstJobRequestRenewed, CancellationToken token)
+        {
+            if (this._isRunServiceJob)
+            {
+                var runServer = await GetRunServerAsync(systemConnection);
+                await RenewJobRequestAsync(runServer, message.Plan.PlanId, message.JobId, firstJobRequestRenewed, token);
+            }
+            else
+            {
+                var runnerServer = HostContext.GetService<IRunnerServer>();
+                await RenewJobRequestAsync(runnerServer, poolId, requestId, lockToken, orchestrationId, firstJobRequestRenewed, token);
+            }
+        }
+
+        private async Task RenewJobRequestAsync(IRunServer runServer, Guid planId, Guid jobId, TaskCompletionSource<int> firstJobRequestRenewed, CancellationToken token)
+        {
+            TaskAgentJobRequest request = null;
+            int firstRenewRetryLimit = 5;
+            int encounteringError = 0;
+
+            // renew lock during job running.
+            // stop renew only if cancellation token for lock renew task been signal or exception still happen after retry.
+            while (!token.IsCancellationRequested)
+            {
+                try
+                {
+                    var renewResponse = await runServer.RenewJobAsync(planId, jobId, token);
+                    Trace.Info($"Successfully renew job {jobId}, job is valid till {renewResponse.LockedUntil}");
+
+                    if (!firstJobRequestRenewed.Task.IsCompleted)
+                    {
+                        // fire first renew succeed event.
+                        firstJobRequestRenewed.TrySetResult(0);
+                    }
+
+                    if (encounteringError > 0)
+                    {
+                        encounteringError = 0;
+                        HostContext.WritePerfCounter("JobRenewRecovered");
+                    }
+
+                    // renew again after 60 sec delay
+                    await HostContext.Delay(TimeSpan.FromSeconds(60), token);
+                }
+                catch (TaskAgentJobNotFoundException)
```

Comment:
does this exception exist in the RunService?

Reply:
ah I thought I am using `TaskAgentJobNotFoundException`, ur right, let me fix it